### PR TITLE
chore(manager): extract some errors as variables

### PR DESCRIPTION
**Pull request by Eoous**
## Overview

Close: #1792 

## Summary by CodeRabbit

- **New Features**
	- Introduced specific error types for block saving operations to enhance clarity in error reporting.

- **Bug Fixes**
	- Improved error handling in the block synchronization process, making it easier to diagnose issues.
	- Enhanced precision in test error handling for block publishing, ensuring accurate error type verification.

### CODE REVIEW

**Comment by kobakaku** [comment]
I think that the error where assert.ErrorContains is used in the test code should have the error string defined separately, and that error string should be used in the test code.
And, there is no need to change the content of the error string.

**Review by kobakaku — COMMENTED** [review]
I think that the error where assert.ErrorContains is used in the test code should have the error string defined separately, and that error string should be used in the test code.
And, there is no need to change the content of the error string.

**Comment by Eoous** [comment]
> I think that the error where assert.ErrorContains is used in the test code should have the error string defined separately, and that error string should be used in the test code. And, there is no need to change the content of the error string.

When error happened, need to return `error` type so the error string would be wrapped as an error type. I think defining error type is better for the reason. And `error` type itself means failed

**Comment by Eoous** [comment]
Request for your @tuxcanfly importent review

**Comment by tuxcanfly** [comment]
I think it makes more sense to have the typed error be a struct like this:

```
type BlockValidationError struct {
	Height uint64
	Err    error
}

func (e BlockValidationError) Error() string {
	return fmt.Sprintf("failed to validate block at height %d: %v", e.Height, e.Err)
}

func (e BlockValidationError) Unwrap() error {
	return e.Err
}

func (m *Manager) trySyncNextBlock(ctx context.Context, daHeight uint64) error {
                ...
                ...
		if err := m.executor.Validate(m.lastState, b); err != nil {
			return BlockValidationError{Height: b.Height(), Err: err}
		}
                ...
                ...
}

```

then the caller can use `errors.As` to get the typed error:


```
if err := m.trySyncNextBlock(ctx, daHeight); err != nil {
    var validationErr BlockValidationError
    if errors.As(err, &validationErr) {
        // Handle validation error specifically
        fmt.Printf("Validation error at height %d: %v\n", validationErr.Height, validationErr.Unwrap())
    } else {
        // Handle other types of errors
        fmt.Printf("Error syncing block: %v\n", err)
    }
}
```

cc @tzdybal

**Comment by Manav-Aggarwal** [comment]
@Eoous Note that this issue was a follow up to https://github.com/rollkit/rollkit/pull/1782#discussion_r1711041710 and this PR was merged to main recently, so this PR should be rebased on top of it and address the follow up issue from #1782.

**Comment by Eoous** [comment]
Maybe we could create new file for defining error structs.

**Comment by Manav-Aggarwal** [comment]
> Maybe we could create new file for defining error structs.

That makes sense, please feel free to do so :)

**Comment by Eoous** [comment]
I will try it this week when have time

**Comment by Manav-Aggarwal** [comment]
Hey @Eoous, note that we recently merged https://github.com/rollkit/rollkit/pull/1789 which contains some significant changes so this PR requires resolving merge conflict and rebasing for re-review.

**Comment by Eoous** [comment]
Solved conflicts and extracted `error types` into errors.go

File: block/errors.go [hunk 1]
```diff
@@ -0,0 +1,41 @@
...
+// These errors are used by Manager.
+var (
+	// ErrNoValidatorsInState is used when no validators/proposers are found in state
+	ErrNoValidatorsInState = errors.New("no validators found in state")
+
+	// ErrNotProposer is used when the manager is not a proposer
+	ErrNotProposer = errors.New("not a proposer")
+)
+
+// SaveBlockError is used when failed to save block
```
Comment (tuxcanfly) [inline block/errors.go:17]:
```suggestion
// SaveBlockError is returned on failure to save block data
```

**Review by tuxcanfly — CHANGES_REQUESTED** [review]

**Comment by Eoous** [comment]
Addressed comments, please review again @tuxcanfly

**Review by tuxcanfly — APPROVED** [review]
OK.

**Review by tzdybal — APPROVED** [review]

**Pull request merged**